File: 2022-04/typescript/94_Binary_Tree_Inorder_Traversal.ts

// When paste to LeetCode comment class TreeNode
class TreeNode {
    val: number
    left: TreeNode | null
    right: TreeNode | null
    constructor(val=0, left=null, right=null) {
        this.val = val
        this.left = left
        this.right = right
    }
}


function inorderTraversal(root: TreeNode): number[] {
    // return inorderTraversalRecursively(root)
    return inorderTraversalIteratively(root)
}


function inorderTraversalRecursively(root: TreeNode): number[] {
    if (!root) return []
    const left = inorderTraversalRecursively(root.left)
    const right = inorderTraversalRecursively(root.right)
    return left.concat(root.val).concat(right)
}


function inorderTraversalIteratively(root: TreeNode): number[] {
    const stack = []
    const result = []
    while (root || stack.length > 0) {
        if (root) {
            stack.push(root)
            root = root.left
        } else {
            const curNode = stack.pop()
            root = curNode.right
            result.push(curNode.val)
        }
    }
    return result
}

File: 2022-04/python/94_Binary_Tree_Inorder_Traversal.py
class Solution:

    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        # return self.inorderTraversalRecursively(root)
        return self.inorderTraversalIteratively(root)

    def inorderTraversalRecursively(self, root: Optional[TreeNode]) -> List[int]:
        if not root: return []
        return self.inorderTraversalRecursively(root.left) + [root.val] + self.inorderTraversalRecursively(root.right)
    
    def inorderTraversalIteratively(self, root: Optional[TreeNode]) -> List[int]:
        result, stack = [], []
        while root or stack:
            if root:
                stack.append(root)
                root = root.left
            else:
                cur_node = stack.pop()
                root = cur_node.right
                result.append(cur_node.val)
        return result